python/nano/test/test_pytorch_onnx.py

```diff
@@ -0,0 +1,110 @@
+#
+# Copyright 2016 The BigDL Authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+from unittest import TestCase
+import pytest
+import os
+import numpy as np
+
+import torch
+from torch.nn import functional as F
+from torch import nn
+from pytorch_lightning.core.lightning import LightningModule
+from bigdl.nano.pytorch.trainer import Trainer
+from torch.optim import Adam
+from torch.utils.data import DataLoader, TensorDataset
+
+from bigdl.nano.pytorch.onnx import onnxruntime
+
+# adaptted from 
+# https://pytorch-lightning.readthedocs.io/en/latest/starter/introduction_guide.html
+# model definition
+@onnxruntime()
```

Comment:
Will we be able to support keras in future?

Reply:
I think so, but I am not sure if it can work in the same manner. https://github.com/onnx/tensorflow-onnx#python-api-reference
Maybe we can 
1. just put the decorator inside pytorch dir and move it out when keras support is ready. 
2. make it `nano.onnxruntime` and give some warning if users use it on a keras.Model now.

Reply:
> I think so, but I am not sure if it can work in the same manner. [onnx/tensorflow-onnx#python-api-reference](https://github.com/onnx/tensorflow-onnx#python-api-reference) Maybe we can
> 
> 1. just put the decorator inside pytorch dir and move it out when keras support is ready.
> 2. make it `nano.onnxruntime` and give some warning if users use it on a keras.Model now.

Make it `nano.onnxruntime` and report error ("tf not currently supported") for now?

Reply:
currently I will just implement `trainer.compile(onnx=True)` and make all our users with `nn.module` to use this API. We believe that is the unified API that can handle all the use cases.

Reply:
Should be `trainer.compile(model, onnx=True)`? Or `nano.pytorch.Trainer.compile(model, onnx=True)`?

Reply:
> Should be `trainer.compile(model, onnx=True)`? Or `nano.pytorch.Trainer.compile(model, onnx=True)`?

Since trainer is not binded to any model instance, there is not much of a difference between these two. a `trainer.compile(model, onnx=True)` is good enough I think. of course, trainer is an instance of `nano.pytorch.Trainer`

Reply:
I think the only difference is if the user needs to first create a `Trainer` instance or not; and `trainer.compile` does not bind `model` to `trainer`, which may be a little bit odd?

Reply:
> I think the only difference is if the user needs to first create a `Trainer` instance or not; and `trainer.compile` does not bind `model` to `trainer`, which may be a little bit odd?

Well, since they need trainer to `fit` or `predict` as long as they want to use our acceleration, I think we are good. btw, `trainer.fit` and other methods in trainer does not bind model to trainer as well.

Reply:
ok

Reply:
oh, I just though about a thing. What if a user has a pl.lightningmodule and he/she just want to use onnx inference support...? It would be odd to use `trainer.compile` to support this case...

Reply:
I think decorator alone is problematic if the class is not defined by the user (e.g., imported from a 3rd party library); you need to use some method to convert either the class or the instance. Not sure if you want to have another method (say `onnxrt`) or just use `compile`

Reply:
> I think decorator alone is problematic if the class is not defined by the user (e.g., imported from a 3rd party library); you need to use some method to convert either the class or the instance. Not sure if you want to have another method (say `onnxrt`) or just use `compile`

The `trainer.compile` thing can definitely handle all the use case when user only have a `nn.module` class, instance or create function. This API will be implemented after the pl module PR.

For the user who have a pl.lightningmodule, I will find a way to support it. My current plan is `trainer.compile(model, onnx=True)`, when the function found that `model` is a pl.lightningmodule, it will ignore the `loss` and `opt` param and just add the onnx support.

The decorator thing will be eliminated.

Reply:
> > I think the only difference is if the user needs to first create a `Trainer` instance or not; and `trainer.compile` does not bind `model` to `trainer`, which may be a little bit odd?
> 
> Well, since they need trainer to `fit` or `predict` as long as they want to use our acceleration, I think we are good. btw, `trainer.fit` and other methods in trainer does not bind model to trainer as well.

One question - if the model is defined in one library (e.g., Chornos) but used in another place, should the one that defines the model to call `compile`, or the one that uses the model to call `compile`? If it's the former, will it be odd that he or she must first define a trainer, or you think it's OK?

Reply:
> 
> 
> > > I think the only difference is if the user needs to first create a `Trainer` instance or not; and `trainer.compile` does not bind `model` to `trainer`, which may be a little bit odd?
> > 
> > 
> > Well, since they need trainer to `fit` or `predict` as long as they want to use our acceleration, I think we are good. btw, `trainer.fit` and other methods in trainer does not bind model to trainer as well.
> 
> One question - if the model is defined in one library (e.g., Chornos) but used in another place, should the one that defines the model to call `compile`, or the one that uses the model to call `compile`? If it's the former, will it be odd that he or she must first define a trainer, or you think it's OK?

**Of course the one who uses the model should call `compile`.** ONNXRuntime support is only an extension on a **model instance**, and the one who defines the model only give the network topology, forwarding (and training loop if valid). The  users who want to use that model will instantiate the model and call `compile` to add the ONNXRuntime support to its model instance.